src/util/crypto_hash.h

```diff
@@ -14,6 +14,23 @@ class crypto_hash
 
   bool operator<(const crypto_hash &h2) const;
 
+  std::array<unsigned int, 5> to_array() const
+  {
+    // TODO: CPP+20 will have a to_array function
+    std::array<unsigned int, 5> result;
+    std::copy(hash, hash + 5, result.begin());
+    return result;
+  }
+
+  size_t to_size_t() const
+  {
+    size_t result = hash[0];
+    for(int i = 1; i < 5; i++)
+      // Do we care about overlaps?
+        result ^= (hash[i] << (sizeof(unsigned int)*i));
```

Comment:
You're throwing away bits of the hash through this shift. As `crypto_hash` is already supposed to be a crypto-strength hash¹, the probability of `hash[0] ^ hash[1]` having any recognizeable pattern is the same as that of `hash[0] ^ (hash[1] << 4)`. Similarly the probability of h0,h1 is the same as that of h1,h0. You're not combining hash values of independent objects where the order matters. I would remove the shift.

¹ I am aware that this is SHA-1, which is broken. Not for hash-maps, though. Still, if that's a problem change the implementation to anything that satisfies the `crypto_hash` promise.

Reply:
I have changed the implementation to use only ^. I will also start another sv-comp run to see if it changes anything.